src/pymor/reductors/dwr.py

```diff
@@ -0,0 +1,260 @@
+            error_estimator = DWRCoerciveRBEstimator(primal_rom.error_estimator, dual_estimators, dual_roms)
+
+        with self.logger.block('Building corrected output ...'):
+            corrected_output = self._build_corrected_output(primal_rom, dual_roms)
+
+        with self.logger.block('Building ROM ...'):
+            rom = primal_rom.with_(output_functional=corrected_output, error_estimator=error_estimator)
+
+        return rom
+
+    def _reduce_to_subbasis(self, primal_dim, dual_dims):
+        primal_rom = self.primal_reductor.reduce(primal_dim)
+        dual_roms = [red.reduce(dim) for red, dim in zip(self.dual_reductors, dual_dims)]
+        corrected_output = self._build_corrected_output(primal_rom, dual_roms, primal_dim, dual_dims)
+        error_estimator = self.assemble_error_estimator_for_subbasis(dual_roms, primal_dim, dual_dims)
+        rom = primal_rom.with_(output_functional=corrected_output, error_estimator=error_estimator)
+        return rom
+
+    def _build_corrected_output(self, primal_rom, dual_roms, primal_dim=None, dual_dims=None):
+        dual_projected_primal_residuals = []
+        dual_dims = dual_dims or [None for _ in range(self.fom.dim_output)]
+        for dual_reductor, dual_dim in zip(self.dual_reductors, dual_dims):
+            dual_basis = dual_reductor.bases['RB'][:dual_dim]
+            op = project(self.fom.operator, dual_basis,
+                         self.primal_reductor.bases['RB'][:primal_dim])
+            rhs = project(self.fom.rhs, dual_basis, None)
+            primal_residual = ResidualOperator(op, rhs, name='dual_projected_residual')
+            dual_projected_primal_residuals.append(primal_residual)
+        return CorrectedOutputFunctional(primal_rom.output_functional, dual_roms,
+                                         dual_projected_primal_residuals)
+
+    @classmethod
+    def create_dual_model(cls, model, dim=0):
+        """Return dual model with the output as right hand side.
+
+        The dual equation is defined as to find the solution p such that
+
+            a(q, p) = - l_dim(q),       for all q,
+
+        where l_dim denotes the dim-th component of the output functional l.
+        See :cite:`Haa17` (Definition 2.26)
+
+        Parameters
+        ----------
+        model
+            The |Model| for which to construct the dual model
+        dim
+            The dimension of the `fom.output_functional` for which the dual model
+            is to be built.
+
+        Returns
+        -------
+        A |Model| with the adjoint operator and the corresponding right-hand side
+        """
+        assert 0 <= dim < model.dim_output
+        output = model.output_functional
+        if isinstance(output, ConcatenationOperator) and isinstance(output.operators[1], BlockColumnOperator):
+            assert isinstance(output.operators[0], NumpyConversionOperator) and len(output.operators) == 2
```

Comment:
Don't fail in case this isn't satisfied but fall back to the generic case. Possibly log a warning.